### PR TITLE
Bootstrap

### PR DESCRIPTION
**Pull request by pythonpete32**
- import relevant packages
- template contracts
- template tests
- cicd

### CODE REVIEW

File: .editorconfig
Comment (Michael-A-Heuer) [inline .editorconfig:1]: Duplicates / can conflict with .vscode/settings.json. In osx-commons, Mathias deleted it.

File: foundry.toml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+[fmt]
+bracket_spacing = true
+int_types = "long"
+line_length = 120
+multiline_func_header = "all"
+number_underscore = "thousands"
+quote_style = "double"
+tab_width = 4
+wrap_comments = true
```
Comment (Michael-A-Heuer) [inline foundry.toml:33]: Duplicates prettier

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+[submodule "lib/osx"]
+	path = lib/osx
+	url = https://github.com/aragon/osx
+[submodule "lib/openzeppelin-contracts-upgradeable"]
+	branch = "release-v4.9"
+	path = lib/openzeppelin-contracts-upgradeable
+	url = https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable
+[submodule "lib/ens-contracts"]
+	path = lib/ens-contracts
+	url = https://github.com/ensdomains/ens-contracts
```
Comment (Michael-A-Heuer) [inline .gitmodules:21]: Probably not needed (we should remove them too on the hardhat template). Feel free to do this in another task.
Generally, let's only include what is strictly needed.

File: .env.example [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+export FORK_BLOCK=18335949
+
+# DAOFactory, RepoFactory
+export MAINNET=0xA03C2182af8eC460D498108C92E8638a580b94d4,0x96E54098317631641703404C06A5afAD89da7373
+export GOERLI=0x1E4350A3c9aFbDbd70FA30B9B2350B9E8182449a,0x477EB3b39C92c38B43778266b09471285e0F7808
+export SEPOLIA=0x7a62da7B56fB3bfCdF70E900787010Bc4c9Ca42e,0x07f49c49Ce2A99CF7C28F66673d406386BDD8Ff4
+export POLYGON=0x51Ead12DEcD31ea75e1046EdFAda14dd639789b8,0x6E924eA5864044D8642385683fFA5AD42FB687f2
+export MUMBAI=0xc715336B5E7F10294F36CA09f19A0493070E2eFB,0x4E7c97ab08c046A8e43571f9839d768ae84492e4
+export BASE_MAINNET=0x1B6632D1357D26055D6452E3b3BEBd1c1d982e86,0xE47C0B1Ea698F2914b6CC218cdC3A5Ca4FAcff18
+export BASE_GOERLI=0x3E773BCb6a8C5ED2f23BF167284DF45D9A550fF6,0x1B6632D1357D26055D6452E3b3BEBd1c1d982e86
```
Comment (Michael-A-Heuer) [inline .env.example:16]: Whitespace at the bottom

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+
+      - name: Install Foundry
+        uses: foundry-rs/foundry-toolchain@v1
+        with:
+          version: nightly
```
Comment (Michael-A-Heuer) [inline .github/workflows/test.yml:23]: Better to use LTS?

File: src/SimpleStorage.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    /// @notice Stores a new number to storage. Caller needs STORE_PERMISSION.
+    /// @param _number The number to be stored.
+    function storeNumber(uint256 _number) external auth(STORE_PERMISSION_ID) {
+        number = _number;
+    }
+
+    /// @notice This empty reserved space is put in place to allow future versions to add new variables without shifting
+    /// down storage in the inheritance chain (see [OpenZepplins guide about storage
+    /// gaps](https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps)).
+    uint256[50] private __gap;
```
Comment (Michael-A-Heuer) [inline src/SimpleStorage.sol:33]:
```suggestion
    uint256[49] private __gap;
```
(number occupies one slot)

File: src/SimpleStorageSetup.sol [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    /// @inheritdoc IPluginSetup
+    function prepareInstallation(
+        address _dao,
+        bytes memory _data
+    ) external returns (address plugin, PreparedSetupData memory preparedSetupData) {
+        uint256 number = abi.decode(_data, (uint256));
+
+        plugin = createERC1967Proxy(
+            IMPLEMEMTATION,
+            abi.encodeWithSelector(SimpleStorage.initialize.selector, _dao, number)
```
Comment (Michael-A-Heuer) [inline src/SimpleStorageSetup.sol:26]: please create a task to update the examples to use [`abi.encodeCall`](https://solidity-by-example.org/abi-encode/) instead `abi.encodeWithSelector`

File: test/SimpleStorage.t.sol [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    function test_store_number() public {
+        vm.prank(address(dao));
+        plugin.storeNumber(69);
+        assertEq(plugin.number(), 69);
+    }
+
+    function test_reverts_if_not_auth() public {
+        // error DaoUnauthorized({dao: address(_dao),  where: _where,  who: _who,permissionId: _permissionId });
+        vm.expectRevert(
+            abi.encodeWithSelector(
```
Comment (Michael-A-Heuer) [inline test/SimpleStorage.t.sol:60]: see above

File: foundry.toml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+cbor_metadata = false
+evm_version = "paris"           # See https://www.evmdiff.com/features?name=PUSH0&kind=opcode
+fuzz = { runs = 1_000 }
+gas_reports = ["*"]
+libs = ["lib"]
+optimizer = true
+optimizer_runs = 10_000
+out = "out"
+script = "script"
+solc = "0.8.21"
```
Comment (Michael-A-Heuer) [inline foundry.toml:16]: By using this solc version, devs won't be able to publish to some rollups (that don't support the PUSH0 opcode yet).

File: src/SimpleStorage.sol [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.8.17;
```
Comment (Michael-A-Heuer) [inline src/SimpleStorage.sol:2]: Match with other pragmas

**Review by Michael-A-Heuer — CHANGES_REQUESTED** [review]
A couple of suggestions. The tests I can review in a second round.

File: .gitmodules (hunk 2 above)
Comment (pythonpete32) [reply .gitmodules:21]: these are actually needed, when you import OSx as a gitsubmodule the ens contracts are not bundled with it

File: src/SimpleStorageSetup.sol (hunk 6 above)
Comment (pythonpete32) [reply src/SimpleStorageSetup.sol:26]: this is Dope!

File: test/SimpleStorage.t.sol (hunk 7 above)
Comment (pythonpete32) [reply test/SimpleStorage.t.sol:60]: but this is a custom error, im not sure that is possible

File: foundry.toml (hunk 8 above)
Comment (pythonpete32) [reply foundry.toml:16]: this was using `0.8.17` originally but you sugested a later version, but i think this was pre arbitrum. regardless will switch back

**Comment by pythonpete32** [comment]
> A couple of suggestions. The tests I can review in a second round.

The tests are placeholders anyway, I will replicate the simple storage tests once this is merged

File: .gitmodules (hunk 2 above)
Comment (Michael-A-Heuer) [inline .gitmodules:21]: Please remove

File: src/SimpleStorage.sol [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+import {PluginUUPSUpgradeable, IDAO} from "@aragon/osx/core/plugin/PluginUUPSUpgradeable.sol";
+
+/**
+ * @title Budget
+ * @author DAOBox (Security@DAOBox.app)
+ * @notice Budgeting module for efficient spending from an Aragon OSx DAO using allowance chains
+ * to delegate spending authority
```
Comment (Michael-A-Heuer) [inline src/SimpleStorage.sol:10]: ?

File: src/SimpleStorage.sol [hunk 11]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.8.17;
```
Comment (Michael-A-Heuer) [inline src/SimpleStorage.sol:2]: Use the same pragma everywhere

File: src/SimpleStorage.sol [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+import {PluginUUPSUpgradeable, IDAO} from "@aragon/osx/core/plugin/PluginUUPSUpgradeable.sol";
+
+/**
+ * @title Budget
+ * @author DAOBox (Security@DAOBox.app)
+ * @notice Budgeting module for efficient spending from an Aragon OSx DAO using allowance chains
+ * to delegate spending authority
+ */
+contract SimpleStorage is PluginUUPSUpgradeable {
```
Comment (Michael-A-Heuer) [inline src/SimpleStorage.sol:12]: Would be good to use the same plugin/plugin naming everywhere https://github.com/aragon/osx-plugin-template-hardhat/blob/main/packages/contracts/src/MyPlugin.sol

**Review by Michael-A-Heuer — CHANGES_REQUESTED** [review]
Minor changes required

**Pull request merged**